onnxruntime/contrib_ops/cpu/quantization/gather_block_quantized.cc

```diff
@@ -0,0 +1,293 @@
+};
+
+template <typename T1, typename Tind>
+Status GatherBlockQuantized<T1, Tind>::PrepareForCompute(OpKernelContext* context, Prepare& p) const {
+  p.data_tensor = context->Input<Tensor>(0);
+  p.indices_tensor = context->Input<Tensor>(1);
+  p.scales_tensor = context->Input<Tensor>(2);
+  p.zero_points_tensor = context->Input<Tensor>(3);
+
+  const auto& data_shape = p.data_tensor->Shape();
+  const auto& indices_shape = p.indices_tensor->Shape();
+  const auto data_rank = data_shape.NumDimensions();
+  p.gather_axis = HandleNegativeAxis(gather_axis_, narrow<int64_t>(data_rank));
+  p.quantize_axis = HandleNegativeAxis(quantize_axis_, narrow<int64_t>(data_rank));
+
+  std::vector<int64_t> shape;
+  shape.reserve(data_rank - 1 + indices_shape.NumDimensions());
+
+  // get output tensor
+  // replace the dimension for p.gather_axis with the shape from the indices
+  for (int64_t i = 0; i < p.gather_axis; ++i)
+    shape.push_back(data_shape[narrow<size_t>(i)]);
+
+  for (const auto dim : indices_shape.GetDims())
+    shape.push_back(dim);
+
+  for (int64_t i = p.gather_axis + 1; i < static_cast<int64_t>(data_rank); ++i)
+    shape.push_back(data_shape[narrow<size_t>(i)]);
+
+  p.output_tensor = context->Output(0, TensorShape(std::move(shape)));
+
+  // validate quantization parameters
+  const auto& scales_shape = p.scales_tensor->Shape();
+  ORT_RETURN_IF_NOT(data_shape.NumDimensions() == scales_shape.NumDimensions(),
+                    "data and scales must have the same rank.");
+  for (size_t i = 0; i < data_shape.NumDimensions(); ++i) {
+    ORT_RETURN_IF_NOT(i == static_cast<size_t>(p.quantize_axis)
+                          ? (data_shape[i] + block_size_ - 1) / block_size_ == scales_shape[i]
+                          : data_shape[i] == scales_shape[i],
+                      "data and scales do not match shapes.");
+  }
+
+  if (p.zero_points_tensor) {
+    const auto& zero_points_shape = p.zero_points_tensor->Shape();
+    ORT_RETURN_IF_NOT(scales_shape.NumDimensions() == zero_points_shape.NumDimensions(),
+                      "scales and zero_points must have the same rank.");
+    for (size_t i = 0; i < scales_shape.NumDimensions(); ++i) {
+      ORT_RETURN_IF_NOT(scales_shape[i] == zero_points_shape[i],
+                        "scales and zero_points must have the same shape.");
+    }
+  }
+
+  // validate indices
+  auto gather_axis_dim = data_shape[narrow<size_t>(p.gather_axis)];
+  auto N = indices_shape.Size();
+  const auto* indices_data = p.indices_tensor->Data<Tind>();
+
+  // TODO(fajin): use SIMD
+  for (int64_t i = 0; i < N; ++i) {
+    auto idx = static_cast<int64_t>(indices_data[i]);
+    if (idx < -gather_axis_dim || idx >= gather_axis_dim) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                             "indices element out of data bounds, idx=", idx,
+                             " must be within the inclusive range [", -gather_axis_dim, ",", gather_axis_dim - 1, "]");
+    }
+  }
```

Comment:
Can we validate this together with the data retrieve.

Reply:
moved to data retrive code in next iter